tests/client.c

```diff
@@ -489,39 +491,82 @@ main(int argc, const char **argv)
       goto cleanup;
     }
 
-    char ech_config_list_buf[10000];
-    size_t ech_config_list_len;
-
-    unsigned int read_result = read_file(getenv("RUSTLS_ECH_CONFIG_LIST"),
-                                         ech_config_list_buf,
-                                         sizeof(ech_config_list_buf),
-                                         &ech_config_list_len);
-    if(read_result != DEMO_OK) {
-      fprintf(stderr,
-              "client: failed to read ECH config list file: '%s'\n",
-              getenv("RUSTLS_ECH_CONFIG_LIST"));
+    // Duplicate the ENV var value - calling STRTOK_R will modify the string
+    // to add null terminators between tokens.
+    char *ech_config_list_copy = strdup(rustls_ech_config_list);
+    if(!ech_config_list_copy) {
+      LOG_SIMPLE("failed to allocate memory for ECH config list");
       goto cleanup;
     }
-    result =
-      rustls_client_config_builder_enable_ech(config_builder,
-                                              (uint8_t *)ech_config_list_buf,
-                                              ech_config_list_len,
-                                              hpke);
-    if(result != RUSTLS_RESULT_OK) {
-      fprintf(stderr, "client: failed to configure ECH");
-      goto cleanup;
+
+    bool ech_configured = false;
+    // Tokenize the ech_config_list_copy by comma. The first invocation takes
+    // ech_config_list_copy. This is reentrant by virtue of saving state to
+    // saveptr. Only the _first_ invocation is given the original string.
+    // Subsequent calls should pass NULL and the same delim/saveptr.
+    const char *delim = ",";
+    char *saveptr = NULL;
+    char *ech_config_list_path =
+      STRTOK_R(ech_config_list_copy, delim, &saveptr);
+
+    while(ech_config_list_path) {
+      // Skip leading spaces
+      while(*ech_config_list_path == ' ') {
+        ech_config_list_path++;
+      }
+
+      // Try to read the token as a file path to an ECH config list.
+      char ech_config_list_buf[10000];
+      size_t ech_config_list_len;
+      const enum demo_result read_result =
+        read_file(ech_config_list_path,
+                  ech_config_list_buf,
+                  sizeof(ech_config_list_buf),
+                  &ech_config_list_len);
+
+      // If we can't read the file, warn and continue
+      if(read_result != DEMO_OK) {
+        // Continue to the next token.
+        LOG("unable to read ECH config list from '%s'", ech_config_list_path);
+        ech_config_list_path = STRTOK_R(NULL, delim, &saveptr);
+        continue;
+      }
+
+      // Try to enable ECH with the config list. This may error if none
+      // of the ECH configs are valid/compatible.
+      result =
+        rustls_client_config_builder_enable_ech(config_builder,
+                                                (uint8_t *)ech_config_list_buf,
+                                                ech_config_list_len,
+                                                hpke);
```

Comment:
I got myself in a bit of a pickle here wondering if `ech_config_list_buf` lives long enough.

`rustls_client_config_builder_enable_ech` says "The caller maintains ownership of the ECH config list TLS bytes", but not really if there is any lifetime requirement for it. From reading the docs I thought there was, but after reading the code I think there isn't.

Reply:
> From reading the docs I thought there was, but after reading the code I think there isn't.

Thanks for calling out that confusion. Presently the docs say:

> The caller maintains ownership of the ECH config list TLS bytes and `rustls_hpke` instance.

What if I amended it to say:

> The caller maintains ownership of the ECH config list TLS bytes and `rustls_hpke` instance. Both may be freed without affecting the `rustls_client_config_builder`.

Reply:
I tacked on a commit that used this language:

> The lifetime of the `ech_config_list_bytes` and `rustls_hpke` instance are not tied to the lifetime of the `rustls_client_config_builder` and can be freed at any time after this function returns without affecting the builder.

Reply:
Maybe "This function does not retain any reference to `ech_config_list_bytes`"? But I think any clarification is an improvement.